File: SmartCard-API-SDK-trial/eGK/eGK.VB/HealthCard_KVK.vb
' --------------------------------------------------------------------------------------------
' MctHealthCard.cs
' CardWerk SmartCard API
' Copyright © 2004-2014 CardWerk Technologies
' --------------------------------------------------------------------------------------------
'HIST
'27NOV2013 MJ protected transaction

Imports System
Imports System.Diagnostics
Imports System.Text

Imports Subsembly.SmartCard

Namespace SampleCode_eGK
    ''' <summary>
    ''' This is a simple example for a card service implementation for the old, memory card
    ''' based german health card - Krankenversicherungskarte (KVK).
    ''' </summary>

    Public Class MctHealthCard
        Inherits SampleCode_eGK.HealthCard
        ''' <summary>
        ''' We overrride the abstract CardName as part of our contract with the base class. 
        ''' </summary>

        Public Overrides ReadOnly Property CardName() As String
            Get
                Return "MCT Versichertenkarte (KVK)"
            End Get
        End Property

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="aSlot"></param>
        ''' <returns></returns>

        Public Function Read(ByVal aSlot As CardTerminalSlot) As CardActivationResult
            MyBase.Clear()

            ' Try to acquire a CardHandle to the inserted card. We try three different
            ' memory card types that are used for german health insurance cards.

            Dim aCard As CardHandle = Nothing
            Dim nActivationResult As CardActivationResult

            aCard = aSlot.AcquireCard(CardTypes.SLE4432, nActivationResult)
            If aCard Is Nothing Then
                aCard = aSlot.AcquireCard(CardTypes.SLE4418, nActivationResult)
            End If
            If aCard Is Nothing Then
                aCard = aSlot.AcquireCard(CardTypes.AT24C02SC, nActivationResult)
            End If
            If aCard Is Nothing Then
                Return nActivationResult
            End If

            '

            Try
                ' Now read all data from the card.
                aSlot.BeginTransaction()

                Dim aResponseAPDU As CardResponseAPDU
                Dim vbData As Byte() = Nothing
                Dim nAppOffset As Integer
                Dim nAppLength As Integer

                ' First try a SELECT APPLICATION. This is the official way to get at the
                ' data of the card. However, this is not supported by all readers.

                Dim vbAppID1 As Byte() = CardHex.ToByteArray("D27600000101")
                Dim vbAppID2 As Byte() = CardHex.ToByteArray("D28000000101")

                Dim aSelectAPDU As New CardCommandAPDU(&H0, &HA4, &H4, &H0, vbAppID1)

                aResponseAPDU = aCard.SendCommand(aSelectAPDU)
                If Not aResponseAPDU.IsSuccessful Then
                    aSelectAPDU.SetData(vbAppID2)
                    aResponseAPDU = aCard.SendCommand(aSelectAPDU)
                End If

                ' If successful, then read the application data without ATR header.

                If aResponseAPDU.IsSuccessful Then
                    aResponseAPDU = aCard.ReadBinary(0, 0)
                    If aResponseAPDU.IsError Then
                        Trace.WriteLine("Failed to read application data!")
                        Return CardActivationResult.UnsupportedCard
                    End If

                    vbData = aResponseAPDU.GetData()
                    nAppOffset = 0
                Else
                    ' If the SELECT APPLICATION failed, then try SELECT FILE.
                    ' This SELECT FILE ensures that the entire memory card is available
                    ' through READ BINARY. It might be, that this command is not supported
                    ' by the driver. In this case we just ignore that failure.

                    aResponseAPDU = aCard.SelectFile(&H3F00)
                    ' Ignore!
                    If Not aResponseAPDU.IsSuccessful Then
                    End If

                    ' Read the entire 256 bytes from the memory card using a simple READ
                    ' BINARY command APDU.

                    aResponseAPDU = aCard.ReadBinary(0, 0)
                    If aResponseAPDU.IsError Then
                        Trace.WriteLine("Failed to read card data!")
                        Return CardActivationResult.UnsupportedCard
                    End If

                    vbData = aResponseAPDU.GetData()

                    ' The returned data should now contain the complete health card data.

                    If _IsAtrHeader(vbData) Then
                        Dim nDirOffset As Integer = vbData(3) And &H7F
                        Dim nDirLength As Integer = 2 + vbData(nDirOffset + 1)

                        nAppOffset = nDirOffset + nDirLength
                    Else
                        nAppOffset = 0
                    End If
                End If

                ' Check whether we have a valid template at the suggested nAppOffset.

                If (nAppOffset + 51) > vbData.Length Then
                    Trace.WriteLine("Not enough template data!")
                    Return CardActivationResult.UnsupportedCard
                End If

                If vbData(nAppOffset) <> &H60 Then
                    Trace.WriteLine("Template not found!")
                    Return CardActivationResult.UnsupportedCard
                End If

                Dim nLenByte As Byte = vbData(nAppOffset + 1)
                If nLenByte = &H81 Then
                    nAppLength = 3 + vbData(nAppOffset + 2)
                ElseIf nLenByte < &H80 Then
                    nAppLength = 2 + nLenByte
                Else
                    Trace.WriteLine("Bad template length format!")
                    Return CardActivationResult.UnsupportedCard
                End If

                If (nAppOffset + nAppLength) > vbData.Length Then
                    Trace.WriteLine("Bad template length!")
                    Return CardActivationResult.UnsupportedCard
                End If

                ' Now try to parse the complete BER encoded application data.

                Dim aRootDO As CardDataObject = CardDataObject.Parse(vbData, nAppOffset, nAppLength, CardDataObjectEncoding.BER)

                Dim aInsuranceNameDO As CardDataObject = aRootDO.Find(&H80, False)
                If aInsuranceNameDO IsNot Nothing Then
                    m_sInsuranceName = _GetString(aInsuranceNameDO)
                End If

                Dim aInsuranceNumberDO As CardDataObject = aRootDO.Find(&H81, False)
                If aInsuranceNumberDO IsNot Nothing Then
                    m_sInsuranceNumber = _GetString(aInsuranceNumberDO)
                End If

                Dim aInsuredNumberDO As CardDataObject = aRootDO.Find(&H82, False)
                If aInsuredNumberDO IsNot Nothing Then
                    m_sInsuredNumber = _GetString(aInsuredNumberDO)
                End If

                Dim aTitleNameDO As CardDataObject = aRootDO.Find(&H84, False)
                If aTitleNameDO IsNot Nothing Then
                    m_sTitleName = _GetString(aTitleNameDO)
                End If

                Dim aFirstNameDO As CardDataObject = aRootDO.Find(&H85, False)
                If aFirstNameDO IsNot Nothing Then
                    m_sFirstName = _GetString(aFirstNameDO)
                End If

                Dim aLastNameDO As CardDataObject = aRootDO.Find(&H87, False)
                If aLastNameDO IsNot Nothing Then
                    m_sLastName = _GetString(aLastNameDO)
                End If

                Dim aStreetDO As CardDataObject = aRootDO.Find(&H89, False)
                If aStreetDO IsNot Nothing Then
                    m_sStreet = _GetString(aStreetDO)
                End If

                Dim aZipCodeDO As CardDataObject = aRootDO.Find(&H8B, False)
                If aZipCodeDO IsNot Nothing Then
                    m_sZipCode = _GetString(aZipCodeDO)
                End If

                Dim aCityCodeDO As CardDataObject = aRootDO.Find(&H8C, False)
                If aCityCodeDO IsNot Nothing Then
                    m_sCity = _GetString(aCityCodeDO)
                End If

                Dim aBirthdayDO As CardDataObject = aRootDO.Find(&H88, False)
                If aBirthdayDO IsNot Nothing Then
                    Dim sBirthday As String = _GetString(aBirthdayDO)
                    m_tBirthday = DateTime.ParseExact(sBirthday, "ddMMyyyy", Nothing)
                End If

                Return CardActivationResult.Success
            Finally
                aSlot.EndTransaction()
                aCard.Dispose()
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="vbData"></param>
        ''' <returns></returns>

        Private Shared Function _IsAtrHeader(ByVal vbData As Byte()) As Boolean
            Debug.Assert(vbData IsNot Nothing)
            Debug.Assert(vbData.Length >= 4)

            Return ((vbData(0) = &H82) OrElse (vbData(0) = &H92) OrElse (vbData(0) = &HA2)) AndAlso (vbData(1) = &H13) AndAlso (vbData(2) = &H10) AndAlso (vbData(3) = &H91)
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="aDO"></param>
        ''' <returns></returns>

        Private Shared Function _GetString(ByVal aDO As CardDataObject) As String
            Dim sb As New StringBuilder(aDO.Length)
            Dim vbValue As Byte() = aDO.Value
            For i As Integer = 0 To vbValue.Length - 1
                Dim b As Byte = vbValue(i)
                If (b < &H20) OrElse (b > &H7F) Then
                    sb.Append("?"c)
                Else
                    sb.Append(g_vchDin66003Table(b - &H20))
                End If
            Next
            Return sb.ToString()
        End Function

        Shared g_vchDin66003Table As Char() = New Char() {" "c, "?"c, "?"c, "?"c, "?"c, "?"c, _
         "&"c, "'"c, "("c, ")"c, "?"c, "+"c, _
         "?"c, "-"c, "."c, "/"c, "0"c, "1"c, _
         "2"c, "3"c, "4"c, "5"c, "6"c, "7"c, _
         "8"c, "9"c, "?"c, "?"c, "?"c, "?"c, _
         "?"c, "?"c, "?"c, "A"c, "B"c, "C"c, _
         "D"c, "E"c, "F"c, "G"c, "H"c, "I"c, _
         "J"c, "K"c, "L"c, "M"c, "N"c, "O"c, _
         "P"c, "Q"c, "R"c, "S"c, "T"c, "U"c, _
         "V"c, "W"c, "X"c, "Y"c, "Z"c, "Ä"c, _
         "Ö"c, "Ü"c, "?"c, "?"c, "?"c, "a"c, _
         "b"c, "c"c, "d"c, "e"c, "f"c, "g"c, _
         "h"c, "i"c, "j"c, "k"c, "l"c, "m"c, _
         "n"c, "o"c, "p"c, "q"c, "r"c, "s"c, _
         "t"c, "u"c, "v"c, "w"c, "x"c, "y"c, _
         "z"c, "ä"c, "ö"c, "ü"c, "ß"c, "?"c}
    End Class
End Namespace

File: SmartCard-API-SDK-trial/eGK/eGK/HealthCard_KVK.cs
// --------------------------------------------------------------------------------------------
// HealthCard_KVK.cs
// CardWerk SmartCard API
// Copyright © 2004-2020 CardWerk Technologies
// 
// CAUTION: 
// OMNIKEY readers with Aviator chipset must be set to 5V power-up voltage
// 
// HIST:
// 11OCT2020 MJ removed CT-API support. OMNIKEY and CHERRY readers are supported via pseudo APDUs implemented on firmware level
// 
// TODO:
// - take card handle as input for generic constructor
// - move eGK, KVK code to CardModule.eGK
// --------------------------------------------------------------------------------------------

using System;
using System.Diagnostics;
using System.Text;

using Subsembly.SmartCard;
using SmartCardAPI.CardModule.MemoryCard;

namespace GermanHealthInsuranceCard
{
    /// <summary>
    /// This is a simple example for a card service implementation for legacy German health card 
    /// This Krankenversicherungskarte (KVK) is based on a memory chip i.e synchronous card that
    /// is not supported by all card readers.
    /// </summary>

    public class HealthCard_KVK : HealthCard
	{
		/// <summary>
		/// We overrride the abstract CardName as part of our contract with the base class. 
		/// </summary>

		public override string CardName
		{
			get
			{
				return "Krankenversichertenkarte KVK";
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSlot"></param>
		/// <returns></returns>

		public CardActivationResult Read(CardTerminalSlot aSlot)
		{
            /// <summary>   Size of KVK data block stored on memory card. </summary>
            const int KVK_DATA_SIZE = 256;

            const string ATR_KVK_SLE4432 = "3B04A2131091";

            base.Clear();

            // Try to acquire a CardHandle to the inserted card. We try three different memory card types that are used for German health insurance cards. All these
            // cards are storage cards with synchronous interface. 
            CardHandle aCard = null; 
			CardActivationResult nActivationResult = CardActivationResult.Unknown;
            aCard = aSlot.AcquireCard(CardTypes.ProcessorCards, out nActivationResult);   // only works for OMNIKEY and CHERRY card readers. For support of other card readers, please use CT-API or contact us for custom implementation 

            if (aCard == null)
            {
                return nActivationResult;
            }

            MemoryCard kvk_memory_card = null;
            string cardName = "SLE4432"; // todo: get name by analyzing the ATR to support the folloowing cards: SLE4432, SLE4442, SLE4418, AT24C02SC
            kvk_memory_card = new MemoryCard(aCard, cardName);

            if(!kvk_memory_card.IsReady)
            {
                return CardActivationResult.UnsupportedCard;
            }

			try
			{
                // Now read all data from the card.

				byte[] vbData = null;
				int nAppOffset;
				int nAppLength;
                
			    vbData = kvk_memory_card.Read(0, KVK_DATA_SIZE);

                // The returned data should now contain the complete health card data.

                if (_IsAtrHeader(vbData))
					{
						int nDirOffset = vbData[3] & 0x7F;
						int nDirLength = 2 + vbData[nDirOffset + 1];

						nAppOffset = nDirOffset + nDirLength;
					}
					else
					{
						nAppOffset = 0;
					}

				// Check whether we have a valid template at the suggested nAppOffset.

				if ((nAppOffset + 51) > vbData.Length)
				{
					Trace.WriteLine("Not enough template data!");
					return CardActivationResult.UnsupportedCard;
				}

				if (vbData[nAppOffset] != 0x60)
				{
					Trace.WriteLine("Template not found!");
					return CardActivationResult.UnsupportedCard;
				}

				byte nLenByte = vbData[nAppOffset + 1];
				if (nLenByte == 0x81)
				{
					nAppLength = 3 + vbData[nAppOffset + 2];
				}
				else if (nLenByte < 0x80)
				{
					nAppLength = 2 + nLenByte;
				}
				else
				{
					Trace.WriteLine("Bad template length format!");
					return CardActivationResult.UnsupportedCard;
				}

				if ((nAppOffset + nAppLength) > vbData.Length)
				{
					Trace.WriteLine("Bad template length!");
					return CardActivationResult.UnsupportedCard;
				}

				// Now try to parse the complete BER encoded application data.

				CardDataObject aRootDO = CardDataObject.Parse(
					vbData, ref nAppOffset, ref nAppLength,
					CardDataObjectEncoding.BER);

				CardDataObject aInsuranceNameDO = aRootDO.Find(0x80, false);
				if (aInsuranceNameDO != null)
				{
					m_sInsuranceName = _GetString(aInsuranceNameDO);
				}

				CardDataObject aInsuranceNumberDO = aRootDO.Find(0x81, false);
				if (aInsuranceNumberDO != null)
				{
					m_sInsuranceNumber = _GetString(aInsuranceNumberDO);
				}

				CardDataObject aInsuredNumberDO = aRootDO.Find(0x82, false);
				if (aInsuredNumberDO != null)
				{
					m_sInsuredNumber = _GetString(aInsuredNumberDO);
				}

				CardDataObject aTitleNameDO = aRootDO.Find(0x84, false);
				if (aTitleNameDO != null)
				{
					m_sTitleName = _GetString(aTitleNameDO);
				}

				CardDataObject aFirstNameDO = aRootDO.Find(0x85, false);
				if (aFirstNameDO != null)
				{
					m_sFirstName = _GetString(aFirstNameDO);
				}

				CardDataObject aLastNameDO = aRootDO.Find(0x87, false);
				if (aLastNameDO != null)
				{
					m_sLastName = _GetString(aLastNameDO);
				}

				CardDataObject aStreetDO = aRootDO.Find(0x89, false);
				if (aStreetDO != null)
				{
					m_sStreet = _GetString(aStreetDO);
				}

				CardDataObject aZipCodeDO = aRootDO.Find(0x8B, false);
				if (aZipCodeDO != null)
				{
					m_sZipCode = _GetString(aZipCodeDO);
				}

				CardDataObject aCityCodeDO = aRootDO.Find(0x8C, false);
				if (aCityCodeDO != null)
				{
					m_sCity = _GetString(aCityCodeDO);
				}

				CardDataObject aBirthdayDO = aRootDO.Find(0x88, false);
				if (aBirthdayDO != null)
				{
					string sBirthday = _GetString(aBirthdayDO);
					m_tBirthday = DateTime.ParseExact(sBirthday, "ddMMyyyy", null);
				}

                
				return CardActivationResult.Success;
			}
			finally
			{
                aCard.Dispose();
			}
		}

		/// <summary>
		/// Memory card contains 4-byte ATR starting at offset 0 
		/// </summary>
		/// <param name="vbData">raw data stored on KVK memory card</param>
		/// <returns></returns>

		static bool _IsAtrHeader(byte[] vbData)
		{
			Debug.Assert(vbData != null);
			Debug.Assert(vbData.Length >= 4);

			return
				((vbData[0] == 0x82) || (vbData[0] == 0x92) || (vbData[0] == 0xA2)) &&
				(vbData[1] == 0x13) && (vbData[2] == 0x10) && (vbData[3] == 0x91);
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="aDO"></param>
		/// <returns></returns>

		static string _GetString(CardDataObject aDO)
		{
			StringBuilder sb = new StringBuilder(aDO.Length);
			byte[] vbValue = aDO.Value;
			for (int i = 0; i < vbValue.Length; ++i)
			{
				byte b = vbValue[i];
				if ((b < 0x20) || (b > 0x7F))
				{
					sb.Append('?');
				}
				else
				{
					sb.Append(g_vchDin66003Table[b - 0x20]);
				}
			}
			return sb.ToString();
		}

		static char[] g_vchDin66003Table = new char[]
		{
			' ', '?', '?', '?', '?', '?', '&', '\'', '(', ')', '?', '+', '?', '-', '.', '/',
			'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '?', '?', '?', '?', '?', '?',
			'?', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
			'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'Ä', 'Ö', 'Ü', '?', '?',
			'?', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
			'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'ä', 'ö', 'ü', 'ß', '?'
		};
	}
}
